### PR TITLE
expression, plan: rewrite builtin function: IF

### PR DESCRIPTION
**Pull request by XuHuaiyu**
#4080 
PTAL @zz-jason @winoros

### CODE REVIEW

File: expression/builtin_control.go [hunk 1]
```diff
@@ -88,44 +87,232 @@ type ifFunctionClass struct {
...
+		ret, isNull = arg1, isNull1
+	}
+	return
+}
+
+type builtinIfDurationSig struct {
+	baseDurationBuiltinFunc
+}
+
+func (b *builtinIfDurationSig) evalDuration(row []types.Datum) (ret types.Duration, isNull bool, err error) {
```
Comment (zz-jason) [inline expression/builtin_control.go]: add these to
```go
_ builtinFunc = &builtinCaseWhenSig{}
```

**Comment by zz-jason** [comment]
reset LGTM

File: expression/builtin_control.go [hunk 2]
```diff
@@ -88,44 +87,232 @@ type ifFunctionClass struct {
...
 
-func (c *ifFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinIfSig{newBaseBuiltinFunc(args, ctx)}
-	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
+// See https://dev.mysql.com/doc/refman/5.7/en/control-flow-functions.html#function_if
+func (c *ifFunctionClass) getFunction(args []Expression, ctx context.Context) (sig builtinFunc, err error) {
+	if err = c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	retTp := c.inferType(args[0].GetType(), args[1].GetType())
```
Comment (lamxTyler) [inline expression/builtin_control.go]: This should be calculated from args0,args1?

File: expression/integration_test.go [hunk 3]
```diff
@@ -1217,6 +1217,17 @@ func (s *testIntegrationSuite) TestControlBuiltin(c *C) {
 	select ifnull(a, 0) as a from t2
 	) t;`)
 	result.Check(testkit.Rows("2.4690"))
+
+	// for if
+	result = tk.MustQuery(`select IF(0,"ERROR","this"),IF(1,"is","ERROR"),IF(NULL,"ERROR","a"),IF(1,2,3)|0,IF(1,2.0,3.0)+0;`)
```
Comment (lamxTyler) [inline expression/integration_test.go]: Add some test that the type of args1 and args2 is different?

**Comment by XuHuaiyu** [comment]
PTAL @lamxTyler

File: plan/typeinfer_test.go [hunk 4]
```diff
@@ -69,18 +69,18 @@ func (s *testPlanSuite) TestInferType(c *C) {
...
-	tests = append(tests, s.createTestCase4Constants()...)
-	tests = append(tests, s.createTestCase4Columns()...)
-	tests = append(tests, s.createTestCase4StrFuncs()...)
-	tests = append(tests, s.createTestCase4MathFuncs()...)
-	tests = append(tests, s.createTestCase4ArithmeticFuncs()...)
-	tests = append(tests, s.createTestCase4LogicalFuncs()...)
+	//tests = append(tests, s.createTestCase4Constants()...)
+	//tests = append(tests, s.createTestCase4Columns()...)
+	//tests = append(tests, s.createTestCase4StrFuncs()...)
+	//tests = append(tests, s.createTestCase4MathFuncs()...)
```
Comment (jackysp) [inline plan/typeinfer_test.go]: Why comment these tests?

**Comment by lamxTyler** [comment]
Seems that there is some thing wrong, `select if(1,1,1.0)` should be:
```
+-------------+
| if(1,1,1.0) |
+-------------+
|         1.0 |
+-------------+

``` 
not
```
+-------------+
| if(1,1,1.0) |
+-------------+
|           1 |
+-------------+
```

**Comment by lamxTyler** [comment]
Another case: `select if(0,1,"1)`, the collation should be `utf8_general_ci`, not `binary`.

File: plan/typeinfer_test.go [hunk 5]
```diff
@@ -417,8 +417,11 @@ func (s *testPlanSuite) createTestCase4LogicalFuncs() []typeInferTestCase {
 
 func (s *testPlanSuite) createTestCase4ControlFuncs() []typeInferTestCase {
 	return []typeInferTestCase{
-		{"ifnull(c_int, c_int)", mysql.TypeLong, charset.CharsetBin, mysql.BinaryFlag, 22, 0},
+		{"ifnull(c_int, c_int    )", mysql.TypeLong, charset.CharsetBin, mysql.BinaryFlag, 22, 0},
 		{"ifnull(c_int, c_decimal)", mysql.TypeNewDecimal, charset.CharsetBin, mysql.BinaryFlag, 17, 3},
+		{"if(c_int, c_decimal, c_int )", mysql.TypeNewDecimal, charset.CharsetBin, mysql.BinaryFlag, 15, 3},
+		{"if(c_int, c_char, c_int    )", mysql.TypeString, charset.CharsetUTF8, 0, 20, -1},
+		{"if(c_int, c_char, c_decimal)", mysql.TypeString, charset.CharsetUTF8, 0, 20, 3},
```
Comment (lamxTyler) [inline plan/typeinfer_test.go]: Add more test here?

**Comment by XuHuaiyu** [comment]
`select if(0,1,"1")` is corresponding to the case `if(c_int, c_char, c_int)`

**Comment by hanfei1991** [comment]
@lamxTyler PTAL

**Comment by XuHuaiyu** [comment]
\cc @hicqu

**Pull request merged**